EIPS/eip-draft-blake2b-f-precompile.md

```diff
@@ -0,0 +1,140 @@
+
+The precompile requires 6 inputs, corresponding to the inputs specified in the [BLAKE2b RFC](https://tools.ietf.org/html/rfc7693#section-3.2)
+
+- `h` - the state vector
+- `m` - the message block vector
+- `t_0, t_1` - offset counters
+- `f` - the final block indicator flag
+- `rounds` - the number of rounds
+
+Loosely packing the data in equal 32-byte slots means an easier job for app developers, and also appears to be compatible with how other precompiles accept input data.
+
+```
+[ bytes32[0] | bytes32[1] ][ bytes32[2] | bytes32[3] | bytes32[4] | bytes32[5] ]
+[      64 bytes for h     ][                    128 bytes for m                ]
+
+[             bytes32[6]             ][             bytes32[7]             ]
+[ 24 bytes padding | 8 bytes for t_0 ][ 24 bytes padding | 8 bytes for t_1 ]
+
+[             bytes32[8]          ][              bytes32[9]              ]
+[ 31 bytes padding | 1 byte for f ][ 28 bytes padding | 4 bytes for rounds]
+```
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `F`.
+
+```solidity
+    function F(bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f, uint32 rounds) public view returns (bytes32[2] memory) {
+      bytes32[2] memory output;
+
+      bytes32 _t0 = bytes32(uint256(t[0]));
+      bytes32 _t1 = bytes32(uint256(t[1]));
+
+      bytes32 _f;
+      if (f) {
+        _f = hex"0000000000000000000000000000000000000000000000000000000000000001";
+      }
+
+      bytes32 _rounds = bytes32(uint256(rounds));
+
+      bytes32[10] memory args = [ h[0], h[1], m[0], m[1], m[2], m[3], _t0, _t1, _f, _rounds ];
+
+      assembly {
+            if iszero(staticcall(not(0), 0x09, args, 0x140, output, 0x40)) {
+                revert(0, 0)
+            }
+        }
+        return output;
+    }
+
+    function callF() public view returns (bytes32[2] memory) {
+      bytes32[2] memory h;
+      h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f1";
+      h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";
+
+      bytes32[4] memory m;
+      m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
+      m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
+      m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
+      m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";
+
+      uint64[2] memory t;
+      t[0] = 18446744073709551552;
+      t[1] = 18446744073709551615;
+
+      bool f = true;
+
+      uint32 rounds = 12;
+
+      return F(h, m, t, f, rounds);
+    }
+```
+
+### Gas costs and benchmarks
+
+Each operation will cost `GFROUND * rounds`.
```

Comment:
Discussing `GFROUND` price...

I've run benchmarks on my Mac with our [initial implementation](https://github.com/keep-network/go-ethereum/pull/4) and took the same approach for setting the price as we did in [EIP-1108](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1108.md). 

```
BenchmarkPrecompiledEcrecover/-Gas=3000-8                  10000            180309 ns/op
```

```
BenchmarkPrecompiledBlake2F/testVectors2bX_0-Gas=1-8             3000000               382 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_3-Gas=1-8             3000000               383 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_70-Gas=1-8            5000000               382 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_140-Gas=1-8           3000000               374 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_230-Gas=1-8           5000000               379 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_300-Gas=1-8           5000000               387 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_370-Gas=1-8           5000000               462 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_440-Gas=1-8           5000000               477 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_510-Gas=1-8           3000000               421 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_580-Gas=1-8           5000000               376 ns/op
```

Let's assume `ecrecover` is perfectly priced. It costs `3000` gas and takes `~180.3` microseconds on my Mac to execute. It is about `16.64` gas per microsecond of `ecrecover` work.

`F` took `402` nanoseconds (`0.402` microseconds) on average, so the entire invocation with `12` rounds (this is what I used in benchmarks) should cost `0.402 * 16.64 = 6.69` gas.

Knowing we had `12` rounds, it is `0.5575` gas per round.

This looks very promising.

I'll add more test vectors to benchmarks and continue the discussion here.

Reply:
Here's my quick benchmarks, where I had set the price to `1 * rounds` (so the vectors had `12` gas each):

```
Name                                     Gascost    Time (ns)    MGas/S    Gasprice for 10MGas/S    Gasprice for ECDSA eq
-------------------------------------  ---------  -----------  --------  -----------------------  -----------------------
PrecompiledEcrecover/                       3000       149931   20.0092                  1499.31               3000
PrecompiledBlake2F/testVectors2bX_0           12          388   30.9278                     3.88                  7.76357
PrecompiledBlake2F/testVectors2bX_3           12          388   30.9278                     3.88                  7.76357
PrecompiledBlake2F/testVectors2bX_70          12          384   31.25                       3.84                  7.68353
PrecompiledBlake2F/testVectors2bX_140         12          398   30.1508                     3.98                  7.96366
PrecompiledBlake2F/testVectors2bX_230         12          394   30.4569                     3.94                  7.88363
PrecompiledBlake2F/testVectors2bX_300         12          401   29.9252                     4.01                  8.02369
PrecompiledBlake2F/testVectors2bX_370         12          392   30.6122                     3.92                  7.84361
PrecompiledBlake2F/testVectors2bX_440         12          400   30                          4                     8.00368
PrecompiledBlake2F/testVectors2bX_510         12          405   29.6296                     4.05                  8.10373
PrecompiledBlake2F/testVectors2bX_580         12          407   29.484                      4.07                  8.14375

```

So, on my machine, for ECDSA equivalence, the cost `12` should really have been around `7` or `8`. So I'd agree that `1` seems reasonable

Reply:
Also, setting it to do `1200` rounds instead of `12` brings down the gas cost -- then the constant should be below `0.5`

```
Name                                     Gascost    Time (ns)    MGas/S    Gasprice for 10MGas/S    Gasprice for ECDSA eq
-------------------------------------  ---------  -----------  --------  -----------------------  -----------------------
PrecompiledEcrecover/                       3000       153527   19.5405                  1535.27                 3000
PrecompiledBlake2F/testVectors2bX_0         1200        26085   46.0035                   260.85                  509.715
PrecompiledBlake2F/testVectors2bX_3         1200        26437   45.3909                   264.37                  516.593
PrecompiledBlake2F/testVectors2bX_70        1200        26746   44.8665                   267.46                  522.631
PrecompiledBlake2F/testVectors2bX_140       1200        26285   45.6534                   262.85                  513.623
PrecompiledBlake2F/testVectors2bX_230       1200        26337   45.5633                   263.37                  514.639
PrecompiledBlake2F/testVectors2bX_300       1200        26565   45.1722                   265.65                  519.094
PrecompiledBlake2F/testVectors2bX_370       1200        26426   45.4098                   264.26                  516.378
PrecompiledBlake2F/testVectors2bX_440       1200        27148   44.2022                   271.48                  530.486
PrecompiledBlake2F/testVectors2bX_510       1200        26768   44.8296                   267.68                  523.061
PrecompiledBlake2F/testVectors2bX_580       1200        26784   44.8029                   267.84                  523.374
```

Columns

* `MGas/S` - Shows what MGas per second was measured on that machine at that time
* `Gasprice for 10MGas/S` shows what the gasprice should have been, in order to reach 10 MGas/second
* `Gasprice for ECDSA eq` shows what the gasprice should have been, in order to have the same cost/cycle as ecRecover

Reply:
And correspondingly; the 'worst case' is with `1` round, where the constant should be closer to `3`. However, in this scenario the call cost would totally overshadow the dynamic cost anyway

```
Name                                     Gascost    Time (ns)    MGas/S    Gasprice for 10MGas/S    Gasprice for ECDSA eq
-------------------------------------  ---------  -----------  --------  -----------------------  -----------------------
PrecompiledEcrecover/                       3000       152882  19.623                    1528.82               3000
PrecompiledBlake2F/testVectors2bX_0            1          137   7.29927                     1.37                  2.68835
PrecompiledBlake2F/testVectors2bX_3            1          132   7.57576                     1.32                  2.59023
PrecompiledBlake2F/testVectors2bX_70           1          142   7.04225                     1.42                  2.78646
PrecompiledBlake2F/testVectors2bX_140          1          146   6.84932                     1.46                  2.86495
PrecompiledBlake2F/testVectors2bX_230          1          142   7.04225                     1.42                  2.78646
PrecompiledBlake2F/testVectors2bX_300          1          146   6.84932                     1.46                  2.86495
PrecompiledBlake2F/testVectors2bX_370          1          146   6.84932                     1.46                  2.86495
PrecompiledBlake2F/testVectors2bX_440          1          138   7.24638                     1.38                  2.70797
PrecompiledBlake2F/testVectors2bX_510          1          143   6.99301                     1.43                  2.80609
PrecompiledBlake2F/testVectors2bX_580          1          144   6.94444                     1.44                  2.82571
```

Reply:
`Gas costs and benchmarks` section has been updated with `1 gas/round` proposition followed by benchmarks.